### ISSUE TITLE
[New Item] Drainable Goo

### ISSUE THREAD

**Issue by Luke18033** [TeamSpen210/BEE2-items#584]
This could be a separate item or integrated into the standard Goo item. In the editor, an item could be connected to it. Then in the game you could activate the input to drain the goo. Deactivating the input would fill the goo again. Start Enabled could be used to make the goo start drained. In BTS or other styles with bottomless pits, it would still use goo pits.

**Comment by TeamSpen210** [comment]
2 issues: 

* Water textures cannot normally move. There's a special version of the material that could be used, but it'll look different (much lower quality).
* I can't change the editor goo behaviour - you won't be able to add portal surfaces, or non-panel items in the goo.

**Comment by TheDarkBomber** [comment]
Bottomless pits would also be a problem.

**Comment by Luke18033** [comment]
@TheDarkBomber I'm pretty sure I just said, *In BTS or other styles with bottomless pits, it would still use goo pits.*

@TeamSpen210 Fixes:

- Use the alternate low-quality texture, or if possible use a custom texture.
- Maybe it wouldn't appear as goo in the editor, meaning it wouldn't appear to "fill up" the chamber, but it would fill the chamber when the map was compiled.

**Comment by TheDarkBomber** [comment]
@Luke18033 

It's based on skyboxes.

**Comment by Luke18033** [comment]
@TheDarkBomber In that case, it would still use goo on *skyboxes* which had bottomless pits.

**Comment by TheDarkBomber** [comment]
But Goo doesn't fit in BTS style.

**Comment by Luke18033** [comment]
@TheDarkBomber There could be a Closable Bottomless Pit which would replace Drainable Goo on skyboxes with bottomless pits

**Comment by TeamSpen210** [comment]
I'll have to implement it as a fully custom item - generating the brushes via the compiler.

**Comment by Luke18033** [comment]
@TeamSpen210 Well, if you think it's doable you can add it.

**Comment by Luke18033** [comment]
@TeamSpen210 At first I thought this was possible because of the part in Aperture Tag where Nigel drains the goo (which was already that way at the start of the map) and you accidentally fill it up again. Anyway, I just figured out that they didn't actually drain/fill the goo. They use worldportals. :unamused:

**Comment by Enterpise** [comment]
So, shouldn't this be closed?

**Comment by TeamSpen210** [comment]
It's possible, you just can't get the fancy water effects (real-time reflection and refraction). Valve also used teleportation for the Escape's blackout scene, since static lighting also is higher quality than switched lights.

**Comment by CrashingScience** [comment]
A large shutter which closes over the bottom of the bottomless pit

**Comment by Luke18033** [comment]
@TeamSpen210 Would it be possible to get the fancy effects when it is filled and *not* moving?

**Comment by TheDarkBomber** [comment]
@CrashingScience 

Wouldn't it close over the top? There is no bottom for the shutter to close over.

**Comment by CrashingScience** [comment]
It should close over the bottom of where the goo would be if it was goo, so
you still get a pit

**Comment by Luke18033** [comment]
@TeamSpen210 Maybe you could make the goo only use the fancy textures when still, and switch to the non-fancy goo when filling/draining.

**Comment by xDadiKx** [comment]
@Luke18033 get on gitter.

**Comment by TheDarkBomber** [comment]
The title doesn't say /Closable Bottomless Pit anymore

**Comment by TeamSpen210** [comment]
In BTS, it'll still be a goo pit. (Where would the shutter go when open?) It'd be far too complex to try to figure out exactly where each shutter should move to, and ensure it doesn't move through another wall.

I can't make the fancy textures switch, those only actually work when the goo is static and immovable.

**Comment by TheDarkBomber** [comment]
It wouldn't work with the style.

**Comment by TeamSpen210** [comment]
It'll be a different texture in BTS, more water-like.

**Comment by TheDarkBomber** [comment]
Would it be useful though? You can't place anything under it.

**Comment by TeamSpen210** [comment]
It'll be a custom item, not the normal goo type. That way the editor doesn't know the area ia covered in goo, and so it'll let you place any item in that pit area.

**Comment by TheDarkBomber** [comment]
So, the ingame model would be glass or fizzler?

**Comment by Luke18033** [comment]
@TeamSpen210 Goo would still generate as usual in-game, even with the custom item, right

**Comment by TeamSpen210** [comment]
The goo would generate normally (aside from being movable), yes.

**Comment by Luke18033** [comment]
@TeamSpen210 One problem, what if you put the entry door in drainable goo set to start enabled? Maybe you could prevent this somehow (only with the entry door though).

**Comment by TeamSpen210** [comment]
I'll want to prevent you from doing that, yes.

**Comment by Luke18033** [comment]
Only the entry door though, not the exit.

**Comment by Luke18033** [comment]
@TeamSpen210 Can't you use the Portal 1 goo texture (which wasn't reflective)

**Comment by TeamSpen210** [comment]
Actually, I converted that to use the `Water` shader as well. It's not really a problem, I just need to make a custom material that's suitable.

**Comment by Luke18033** [comment]
I think there's a non-fancy version of the PeTI goo texture you could use for this, at least in Clean.

**Comment by Endermage77** [comment]
For the bottomless pit shutters, you could do a "Cascading Door" shutter, at the seam of the "goo" and the pit, which when folded up would only occupy 1 block of space.

**Comment by TeamSpen210** [comment]
Well yes, but how does that work for oddly-shaped pits, or ones the size of the whole map?

**Comment by Endermage77** [comment]
Oh yeah, odd-shaped pits would be a bit of an issue to code for...

But in terms of massive pits, simply add additional segments. Aperture is the leading innovator in paper-thin steel doors.

**Comment by Luke18033** [comment]
For the editor model, we could use a recreation of the Goo Tap seen in beta videos (the texture remains in the final, but the model has been removed).

**Comment by CrashingScience** [comment]
A big sliding door that moves quickly from side of the pit to the other. For oddly shaped pits it would just clip into the floor. And for bottomless pits you shouldnt be building things under them, so its not visble when retracted.

**Comment by Luke18033** [comment]
What if you did this?
![20170423102204_1](https://cloud.githubusercontent.com/assets/19228201/25315761/540c7ff0-280f-11e7-8e3a-589c6003270e.jpg)

**Comment by CrashingScience** [comment]
Then deepen the pit to below the rest of the map and only place 'goo' at the very bottom.

**Comment by Luke18033** [comment]
How about just make it use goo?

**Comment by Luke18033** [comment]
Or a "neurotoxin pit", like in the Aperture Tag Stage maps.

**Comment by Endermage77** [comment]
If it were to be a sliding door sliding into the thin sides in that picture above, it could still be done with a collapsing door, I think. It would all just slide into the block space on either side.
Or it could automatically try to slide away into the other sides that aren't one block thick.

If you somehow needed a pit like that that was surrounded by lower bits on all 4 sides, you could perhaps just do a custom Closed Solid Field instead of a door.

**Comment by Luke18033** [comment]
Another problem is if you put items on the bottom. Those items would have to be parented to the door, and would clip through the wall when it was opening or closing. If the CSFi was used, they would end up floating.

**Comment by Luke18033** [comment]
I tried making a drainable goo texture. This is the best I was able to do with LightmappedGeneric.
![image](https://user-images.githubusercontent.com/19228201/89691991-45d28880-d8bf-11ea-9b05-573958e2091a.png)
It looks okay in this screenshot, but the problem is that none of the flowing effects are there - it just scrolls in a fixed direction. Maybe `SolidEnergy` could somehow be used instead, since that supports flowmaps?

**Comment by TeamSpen210** [comment]
Note that the Water shader itself should still be movable, it's just that it can't use the realtime cameras. You'll want to use `func_water_analog`, it's a special version of `func_movelinear` which applies water physics to the brush. That's what HL2's water uses.

**Comment by Luke18033** [comment]
Yeah, I'm using that entity in the screenshot. In my testing, applying a `Water` material - even one without realtime reflection or refraction - to a brush entity would cause that face to appear at the map origin and not move. Unless there's some specific parameter that needs to be set, I don't think that works.

**Comment by Luke18033** [comment]
Hammer:
![image](https://user-images.githubusercontent.com/19228201/89693543-9946d580-d8c3-11ea-8fa7-d075e3f45ffe.png)
In-game:
![image](https://user-images.githubusercontent.com/19228201/89693537-92b85e00-d8c3-11ea-95af-aa67af65012e.png)
The texture being used is `toxicslime002a_noreflect`. It seems that the face is always positioned relative to the world origin, rather than the entity's origin. The actual swimmable area is in the correct location.
